net.go

```diff
@@ -635,6 +657,11 @@ func (m *Memberlist) handleAlive(buf []byte, from net.Addr) {
 		live.Port = uint16(m.config.BindPort)
 	}
 
+	if err := m.ensureCanConnect(from); err != nil {
```

Comment:
Since you are using socket information (e.g. `from net.Addr`) provided in the calling arguments, I think this `ensureCanConnect` should happen as the first line in this method.

Reply:
Edge case:
There is a subtle difference between the socket info `from net.Addr` (used here) and `(*alive).Addr` (used in `aliveNode()`). The former comes straight from our side of the socket firing data into the memberlist implementation and the latter is addressing information sent _over_ that socket from the other side. In almost all cases you would hope that these are the same.

In one very new case (https://github.com/hashicorp/consul/pull/6884) these can differ. The socket address will be functionally useless as it can come from a mesh gateway envoy local to your datacenter, but the `alive{}` structure will contain the original address as the sender views themselves.

So to be absolutely correct here, you should run *both* the socket address *and* the `(*alive).Addr` address through your `ensureCanConnect` function and block it if either fails.

Reply:
@rboyer In `handleAlive()`, `aliveNode()` is evently called in and already contains `m.config.IPAllowed(a.Addr)` ... do you want me to add another check anyway in `handleAlive()` anyway ?

Reply:
Oh right. I guess we can skip the double check in this function given that the latter checks will cover it.